lib/codecs/src/encoding/format/gelf.rs

```diff
@@ -0,0 +1,470 @@
+
+    /// Valides rules for additional field names and value types.
+    fn validate_additional_field(
+        &self,
+        field: &str,
+        value: &Value,
+        log: &LogEvent,
+        conformed_log: &mut Option<LogEvent>,
+    ) -> vector_core::Result<()> {
+        // Additional fields must be prefixed with underscores.
+        // Prepending the underscore since vector adds fields such as 'source_type'
+        // which would otherwise throw errors.
+        if !field.is_empty() && !field.starts_with('_') {
+            self.conformed_log_mut(log, conformed_log)
+                .rename_key(field, &*format!("_{}", &field));
+        }
+
+        // additional fields must be only word chars, dashes and periods.
+        if !VALID_FIELD_REGEX.is_match(field) {
+            return MissingFieldSnafu { field }
+                .fail()
+                .map_err(|e| e.to_string().into());
+        }
+
+        // additional field values must be only strings or numbers
+        if !(value.is_integer() || value.is_float() || value.is_bytes()) {
+            self.err_invalid_type(field, "string or number", value.kind_str())?;
+        }
+        Ok(())
+    }
+
+    /// Validates rules for field names and value types.
+    fn validate_field_names_and_values(
+        &self,
+        log: &LogEvent,
+        conformed_log: &mut Option<LogEvent>,
+    ) -> vector_core::Result<()> {
+        if let Some(event_data) = log.as_map() {
+            for (key, value) in event_data {
+                match key.as_str() {
+                    VERSION | HOST | SHORT_MESSAGE | FULL_MESSAGE | FACILITY | FILE => {
+                        if !value.is_bytes() {
+                            self.err_invalid_type(key, "UTF-8 string", value.kind_str())?;
+                        }
+                    }
+                    TIMESTAMP => {
+                        if !(value.is_timestamp() || value.is_integer()) {
+                            self.err_invalid_type(key, "timestamp or integer", value.kind_str())?;
+                        }
+                    }
+                    LEVEL => {
+                        if !value.is_integer() {
+                            self.err_invalid_type(key, "integer", value.kind_str())?;
+                        }
+                    }
+                    LINE => {
+                        if !(value.is_float() || value.is_integer()) {
+                            self.err_invalid_type(key, "number", value.kind_str())?;
+                        }
+                    }
+                    _ => {
+                        self.validate_additional_field(key, value, log, conformed_log)?;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Determine if input log event is in valid GELF format
+    /// Potentially return a copy of the log event that was modified to conform to valid GELF
+    fn to_gelf_event(&self, log: &LogEvent) -> vector_core::Result<Option<LogEvent>> {
```

Comment:
Apologies for harping on this, but I still find this interface a bit confusing 😅  It's definitely improved from storing the mutated log event on `self` but, what I would expect is for this function to take a `LogEvent` and then return that `LogEvent` (potentially modified). That is the signature would be:

```
    fn to_gelf_event(&self, log: LogEvent) -> vector_core::Result<LogEvent> {
```

Passing the owned log event around would avoid the need to have a separate `conformed_log` argument for the helper functions. The below would then become:

```
let log = self.coerce_required_fields(log)?;
let log = self.coerce_field_names_and_values(log)?;
```

(or you could chain them together with `and_then`: `self.coerce_required_fields(log).and_then(|log| self.coerce_field_names_and_values(log)`)

Does that make sense? I'm curious what you think.

Reply:
Harping is good and appreciated! - as I'm still developing my Rust skills.

I believe what led me to this, is I was trying to avoid was making a copy of the incoming event from `encode()` every time, because we don't always have to do that.

The approach you suggest does make sense, and I'm happy to do it, was just trying to optimize with less copies. But maybe there is a way to achieve that without copying... I'll take another look.

Reply:
Totally. That makes sense to avoid the clone. I believe my suggestion here to take the owned event should avoid the clone as well by just (possibly) mutating it and then returning it back. I could be missing something though.

Reply:
Ok I'll take a closer look at it. I'm doing an MVP test here and not getting it to work due to ownership rules but I might also be missing something.

Reply:
Ok I've given this another look and, it seems if I use the function signature you outlined:

>     fn to_gelf_event(&self, log: LogEvent) -> vector_core::Result<LogEvent> {

, The only way to make that work is to call it by:

```Rust
to_gelf_event(log.to_owned())
```

, and unless I'm mistaken, that is doing a copy.

Reply:
I think that is because you are calling `.as_log()` above at the call site, which turns an `Event` into an `&LogEvent`. Rather than `.as_log()` you can use `.into_log()` to convert the `Event` into an owned `LogEvent`. It looks like all of the places where `to_gelf_event` is called receive an owned `Event` so it seemingly should be possible.

Reply:
Started to look into this- the `.into_log()` is enlightening :)

If `into_log()` converts to an owned LogEvent , then is there any reason we couldn't do:

```Rust
let mut log = event.into_log();
to_gelf_event(&mut log)?;
```

(and the other helper methods receive a `log : &mut LogEvent` , as parameter ?

Reply:
That should work too. My personal preference is still for passing owned values in and returning owned values out rather than passing in a mutable reference `&mut` but I don't feel super strongly about it. Either way is a significant improvement over the current state in my opinion.

Reply:
I think I resolved it in the latest commit.
Had to extract the modification of the additional field keys missing underscore prefix, as was not able to modify the log event while already accessing the map.